File: array/ts/build_array_from_permutation.ts
// 1920 基于排列构建数组
// https://leetcode.cn/problems/build-array-from-permutation/
// 预分配等长数组，按下标依次填 ans[i] = nums[nums[i]]，一次遍历 O(n)。
function buildArray(nums: number[]): number[] {
    const n = nums.length;
    // 预分配成长度 n（等价于 Go 的 make([]int, len(nums))）
    const ans: number[] = new Array(n);

    // 取 nums[i] 作为下标再去 nums 里取值，注意是 nums[nums[i]] 不是 nums[i]
    for (let i = 0; i < n; i++) {
        ans[i] = nums[nums[i]];
    }

    // 少写这行 return 会返回 undefined，tsc 在 --strict 下也会直接报 TS2355
    return ans;
}

File: array/python/build_array_from_permutation.py
# 1920. 基于排列构建数组 (Build Array from Permutation)
# https://leetcode.cn/problems/build-array-from-permutation/description/
# 思路：预分配等长数组，逐个按下标赋值 ans[i] = nums[nums[i]]，一次遍历 O(n)。
from typing import List


class Solution:
    def buildArray(self, nums: List[int]) -> List[int]:
        # 必须预分配成长度 len(nums)（或改用 append），空列表不能用下标赋值
        ans = [0] * len(nums)

        # range(len(nums)) 覆盖全部下标；写成 len(nums) - 1 会漏掉最后一个元素
        for i in range(len(nums)):
            ans[i] = nums[nums[i]]

        return ans
